### ISSUE TITLE
Adding spring-boot-starter-hateoas to spring native project results in java.lang.NoClassDefFoundError: javax/servlet/http/HttpServletRequest

### ISSUE THREAD

**Issue by jeffreyhilaire** [spring-projects-experimental/spring-native#1281]
Hi,

I am trying to configure a spring web flux application with the intent of building a native image later on. So far, I am able to build a dummy application into an image using spring native, run it and get a hello world response without issues. 

As soon as I add spring-boot-starter-hateoas ( while excluding spring-boot-starter-tomcat because I don't want to use Tomcat but rather Netty with functional routes and handlers ), I get the error below during the image build process : 

I am not quite sure how to get rid of all the "servlet" dependencies from spring-hateoas if this is at all possible.

```
[INFO] Generating class file bytes for a proxy named org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController$$SpringProxy$a0f480bd
[ERROR] java.lang.NoClassDefFoundError: javax/servlet/http/HttpServletRequest
[ERROR] [java.base/java.lang.Class.getDeclaredMethods0(Native Method), java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166), java.base/java.lang.Class.getDeclaredMethods(Class.java:2309), org.springframework.aop.framework.ProxyGenerator.doValidateClass(ProxyGenerator.java:136), org.springframewo
rk.aop.framework.ProxyGenerator.validateClassIfNecessary(ProxyGenerator.java:123), org.springframework.aop.framework.ProxyGenerator.getProxyBytes(ProxyGenerator.java:72), org.springframework.aot.nativex.ConfigurationContributor.generateBuildTimeClassProxy(ConfigurationContributor.java:182), org.springframework.
aot.nativex.ConfigurationContributor.generateBuildTimeClassProxies(ConfigurationContributor.java:160), org.springframework.aot.nativex.ConfigurationContributor.processBuildTimeClassProxyRequests(ConfigurationContributor.java:137), org.springframework.aot.nativex.ConfigurationContributor.contribute(Configuration
Contributor.java:72), org.springframework.aot.BootstrapCodeGenerator.generate(BootstrapCodeGenerator.java:77), org.springframework.aot.maven.TestGenerateMojo.execute(TestGenerateMojo.java:65), org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137), org.apache.maven.lif
ecycle.internal.MojoExecutor.execute(MojoExecutor.java:210), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148), org.apache.maven.lifecycle.internal.MojoExecutor.executeForkedExecutions(MojoExecutor.jav
a:355), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:200), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148), org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.bu
ildProject(LifecycleModuleBuilder.java:117), org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81), org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56), org.apache.maven.lifecycle.internal.Lifecyc
leStarter.execute(LifecycleStarter.java:128), org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305), org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192), org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105), org.apache.maven.cli.MavenCli.execute(MavenCli.java:972), org.apache.maven.c
li.MavenCli.doMain(MavenCli.java:293), org.apache.maven.cli.MavenCli.main(MavenCli.java:196), java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method), java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62), java.base/jdk.internal.reflect.Delegatin
gMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43), java.base/java.lang.reflect.Method.invoke(Method.java:566), org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282), org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225), org.codehaus.plexus.clas
sworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406), org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)]
```

**Comment by jeffreyhilaire** [comment]
I have just read this statement from the docs https://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#web.spring-hateoas indicating that spring-boot-starter-hateoas is not compatible with spring-native so I switched to including only spring-hateoas but the end result is the same. As long as I don't include Tomcat into the mix, I get a class not found error related to javax/servlet/http/HttpServletRequest

[INFO] Generating class file bytes for a proxy named org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController$$SpringProxy$a0f480bd
[ERROR] java.lang.NoClassDefFoundError: javax/servlet/http/HttpServletRequest
[ERROR] [java.base/java.lang.Class.getDeclaredMethods0(Native Method), java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166), java.base/java.lang.Class.getDeclaredMethods(Class.java:2309), org.springframework.aop.framework.ProxyGenerator.doValidateClass(ProxyGenerator.java:136), org.springframewo
rk.aop.framework.ProxyGenerator.validateClassIfNecessary(ProxyGenerator.java:123), org.springframework.aop.framework.ProxyGenerator.getProxyBytes(ProxyGenerator.java:72), org.springframework.aot.nativex.ConfigurationContributor.generateBuildTimeClassProxy(ConfigurationContributor.java:182), org.springframework.
aot.nativex.ConfigurationContributor.generateBuildTimeClassProxies(ConfigurationContributor.java:160), org.springframework.aot.nativex.ConfigurationContributor.processBuildTimeClassProxyRequests(ConfigurationContributor.java:137), org.springframework.aot.nativex.ConfigurationContributor.contribute(Configuration
Contributor.java:72), org.springframework.aot.BootstrapCodeGenerator.generate(BootstrapCodeGenerator.java:77), org.springframework.aot.maven.TestGenerateMojo.execute(TestGenerateMojo.java:65), org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137), org.apache.maven.lif
ecycle.internal.MojoExecutor.execute(MojoExecutor.java:210), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148), org.apache.maven.lifecycle.internal.MojoExecutor.executeForkedExecutions(MojoExecutor.jav
a:355), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:200), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156), org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148), org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.bu
ildProject(LifecycleModuleBuilder.java:117), org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81), org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56), org.apache.maven.lifecycle.internal.Lifecyc
leStarter.execute(LifecycleStarter.java:128), org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305), org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192), org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105), org.apache.maven.cli.MavenCli.execute(MavenCli.java:972), org.apache.maven.c
li.MavenCli.doMain(MavenCli.java:293), org.apache.maven.cli.MavenCli.main(MavenCli.java:196), java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method), java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62), java.base/jdk.internal.reflect.Delegatin
gMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43), java.base/java.lang.reflect.Method.invoke(Method.java:566), org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282), org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225), org.codehaus.plexus.clas
sworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406), org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE

**Comment by snicoll** [comment]
Unfortunately, Spring HATEOAS has no WebFlux support as far as I am aware (see https://github.com/spring-projects/spring-boot/issues/16020#issue-413499902). I am going to close this as it doesn't seem Spring Native specific.

**Comment by jeffreyhilaire** [comment]
@snicoll 

I must admit, it's very confusing. On one hand you say spring hateoas is not supported with webflux, on the other hand spring-hateoas lists example of using it with webflux here https://github.com/spring-projects/spring-hateoas/blob/main/src/test/java/org/springframework/hateoas/support/WebFluxEmployeeController.java.

Also my question is about the conflict between spring-native and spring-hateoas, but it looks like there is much confusion about all of this and I will have to roll my own solution or find something else. All I care about at this point was not re-inventing the wheel and having to re-create all the support about resource representation that exists in spring-hateoas.

Oh well !

**Comment by snicoll** [comment]
> On one hand you say spring hateoas is not supported with webflux, on the other hand spring-hateoas lists example of using it with webflux here

Yes. You mentioned "spring-boot-starter-hateoas" which is the Spring Boot integration of Spring HATEOAS. The link in my previous comment should hopefully clarify that such support doesn't exist. You adding this starter brings Tomcat and WebMVC so that webflux application isn't one really ([read the note in the section of the doc](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#web.reactive.webflux)).


> Also my question is about the conflict between spring-native and spring-hateoas,

I read it more as you trying to use Spring HATEOAS without the Servlet API on the classpath. If you can share a small sample that works on the JVM and fails in Native we can have a look (as a zip attached to this issue or a link to a GitHub repo).

**Comment by jeffreyhilaire** [comment]
@snicoll I appreciate your looking into this, I have posted on Github where I am unable to build a native image as soon as I add spring-hateoas on a branch named "native-build-broken-with-hateoas" (https://github.com/jeffreyhilaire/hateoas-webflux-demo/tree/native-build-broken-with-hateoas)

You will see that the "main" branch (https://github.com/jeffreyhilaire/hateoas-webflux-demo/tree/main) build and runs fine both on the JVM and as a docker image. 

Once I add spring-hateoas ( not spring-boot-starter-hateoas) , the native image build fails with "java.lang.ClassNotFoundException: javax.servlet.http.HttpServletRequest". I have tried many ways to satisfy this dependency without success. It's as if one cannot use spring-hateoas without pulling in a servlet container at all.
